File: fir/reduce_cost.py
import numpy as np
from scipy import interpolate
from scipy.io import loadmat, savemat
from scipy.fft import ifft, ifftshift
import pandas as pd

import matplotlib.pyplot as plt

# USER PARAMETERS
frf_csv = './fir/data/predicted_G_values.csv'  # FRF in CSV: [omega, ReG, ImG]
io_file = './fir/data/data_hour.mat'           # recorded I/O to replay
lambda_factor = 0.995               # RLS forgetting factor
energy_cut = 0.99                   # keep ≥99% of |g| energy
plot_rate = 100                     # samples between plot refreshes

# 1) Load FRF from CSV
df = pd.read_csv(frf_csv)
M = df.values.T  # Transpose to get features as rows
omega = M[0, :]
ReG = M[1, :]
ImG = M[2, :]
G_pos = ReG + 1j * ImG

# Uniform frequency grid
Npos = len(omega)
omega_min = min(omega)
omega_max = max(omega)
Nfft = 2 ** int(np.ceil(np.log2(4 * Npos)))
omega_uni = np.linspace(omega_min, omega_max, Nfft // 2 + 1)

# Interpolation
G_uni = interpolate.pchip_interpolate(omega, G_pos, omega_uni)

# Hermitian symmetric spectrum
G_full = np.concatenate([np.conj(G_uni[Nfft//2-1:0:-1]), G_uni])

# 2) Impulse response via IFFT
g_full = np.real(ifft(ifftshift(G_full)))

# Sampling frequency
Dw = omega_uni[1] - omega_uni[0]
Fs = Dw * Nfft / (2 * np.pi)
Ts = 1 / Fs

# Trim by energy
Etotal = np.sum(np.abs(g_full) ** 2)
cumE = np.cumsum(np.abs(g_full) ** 2)
L = np.where(cumE / Etotal >= energy_cut)[0][0]
L = max(L, 4)

# Apply window
w = np.hanning(L)
h_init = g_full[:L] * w

print(f'[INFO] FIR length L = {L}  (Ts = {Ts:.4g} s)')

# 3) Load I/O data to replay
mat_data = loadmat(io_file)
var_names = [name for name in mat_data.keys() if not name.startswith('__')]
mat = mat_data[var_names[0]]

if mat.shape[0] >= 3 and mat.shape[1] > 3:
    # 3 rows: time; y; u
    time = mat[0, :].T
    y = mat[1, :].T
    u = mat[2, :].T
elif mat.shape[1] >= 3:
    # 3 columns: [time, y, u]
    time = mat[:, 0]
    y = mat[:, 1]
    u = mat[:, 2]
else:
    raise ValueError('I/O data has unexpected size')

# Calculate dt
if len(time) > 1:
    dtv = np.diff(time)
    dt = np.mean(dtv)
    if np.max(np.abs(dtv - dt)) > 0.01 * dt:
        print('Warning: Non-uniform time steps detected.')
else:
    dt = Ts
print(f'Loaded I/O: {len(time)} samples, dt = {dt:.4g} s')

# Resample if necessary
if abs(dt - Ts) > 1e-6:
    print('Resampling to FIR Ts...')
    t_old = time
    t_new = np.arange(0, np.ceil(time[-1] / Ts) + 1) * Ts
    u = np.interp(t_new, t_old, u)
    y = np.interp(t_new, t_old, y)
    dt = Ts
N = len(u)

# 4) Partial‐Update LMS initialization
h = h_init.copy()  # FIR taps
mu = 1e-3          # LMS stepsize
M_update = 10      # number of taps to update per sample
phi = np.zeros(L)

yhat = np.zeros(N)
err = np.zeros(N)
H = np.zeros((L, N))  # Store coefficient history

# 5) Real‐time loop with M‐Max partial‐update LMS
plt.figure(figsize=(10, 8))
plt.suptitle('Partial-Update LMS Identification')

ax1 = plt.subplot(3, 1, 1)
plt.grid(True)
h_meas, = plt.plot([], [], 'k', label='Measured')
h_pred, = plt.plot([], [], 'r--', label='Predicted')
plt.legend()
plt.xlabel('n')
plt.ylabel('y')

ax2 = plt.subplot(3, 1, 2)
plt.grid(True)
h_err, = plt.plot([], [], 'b')
plt.xlabel('n')
plt.ylabel('error')

ax3 = plt.subplot(3, 1, 3)
h_coef, = plt.plot(range(1, L+1), h, 'b-o', markersize=3)
plt.grid(True)
plt.xlabel('Tap index')
plt.ylabel('h')
ax3.set_title('Iteration 0')
plt.ylim([min(h_init)*1.1, max(h_init)*1.1])

plt.tight_layout()

for n in range(N):
    # update input vector
    phi = np.roll(phi, 1)
    phi[0] = u[n]
    
    # prediction and error
    yhat[n] = np.dot(phi, h)
    err[n] = y[n] - yhat[n]
    
    # partial‐update LMS
    delta = mu * phi * err[n]
    idx_sort = np.argsort(np.abs(delta))[::-1]  # Descending order
    S = idx_sort[:min(M_update, L)]
    h[S] += delta[S]
    
    # store coefficient history
    H[:, n] = h
    
    # update plots periodically
    if (n % plot_rate == 0) or (n == N - 1):
        n_plot = n + 1
        h_meas.set_data(range(1, n_plot+1), y[:n_plot])
        h_pred.set_data(range(1, n_plot+1), yhat[:n_plot])
        h_err.set_data(range(1, n_plot+1), err[:n_plot])
        h_coef.set_ydata(h)
        ax3.set_title(f'Iteration {n}')
        
        # Update axis limits
        ax1.relim()
        ax1.autoscale_view()
        ax2.relim()
        ax2.autoscale_view()
        
        plt.pause(0.01)

# 6) Error metrics & save
rmse = np.sqrt(np.mean(err[L:] ** 2))
yn = y[L:] - np.mean(y[L:])
nrmse = 1 - np.linalg.norm(err[L:]) / np.linalg.norm(yn)
R2 = 1 - np.sum(err[L:] ** 2) / np.sum(yn ** 2)

print('\n===== FINAL ERROR =====')
print(f'RMSE  = {rmse:.4g}')
print(f'NRMSE = {nrmse*100:.2f} %')
print(f'R^2   = {R2:.3f}')

# Save results
savemat('pu_lms_results.mat', {
    'h': h,
    'rmse': rmse,
    'nrmse': nrmse,
    'R2': R2,
    'yhat': yhat,
    'err': err,
    'Ts': Ts
})

plt.show()

File: adhoc/dust_code/reduce_cost.m
%% fir_rt_identification_csv.m
% Real-time identification of a flexible link using an FIR model,
% FRF from CSV, I/O from .mat, end with savemat.

clearvars; close all; clc;

%% USER PARAMETERS
frf_csv      = 'predicted_G_values.csv';  % FRF in CSV: [omega, ReG, ImG]
io_file      = 'data_hour.mat';                     % recorded I/O to replay
lambda       = 0.995;                               % RLS forgetting factor
energy_cut   = 0.99;                                % keep ≥99% of |g| energy
plot_rate    = 100;                                 % samples between plot refreshes

%% 1) Load FRF from CSV
% 1行目がヘッダの場合は 'NumHeaderLines',1 を追加
M = readmatrix(frf_csv);           
omega = M(:,1);
ReG   = M(:,2);
ImG   = M(:,3);
G_pos = ReG + 1j*ImG;

% 均一周波数グリッド
Npos      = numel(omega);
omega_min = min(omega);
omega_max = max(omega);
Nfft      = 2^nextpow2(4*Npos);
omega_uni = linspace(omega_min,omega_max,Nfft/2+1).';

% 補間
G_uni = interp1(omega,G_pos,omega_uni,'pchip',0);

% Hermitian 対称スペクトルの生成
G_full = [ conj(G_uni(end-1:-1:2)); G_uni ];

%% 2) インパルス応答 via IFFT
g_full = real( ifft( ifftshift(G_full) ) );

% サンプリング周波数
Dw = omega_uni(2)-omega_uni(1);
Fs = Dw*Nfft/(2*pi);
Ts = 1/Fs;

% エネルギーでトリミング
Etotal = sum(abs(g_full).^2);
cumE   = cumsum(abs(g_full).^2);
L      = find(cumE/Etotal>=energy_cut,1,'first');
L      = max(L,4);

% 窓掛け
w      = hann(L);
h_init = g_full(1:L).*w;

fprintf('[INFO] FIR length L = %d  (Ts = %.4g s)\n',L,Ts);

%% 3) Load I/O data to replay
S = load(io_file);
fn = fieldnames(S);
% __ で始まるものを除外
fn = fn(~startsWith(fn,'__'));
mat = S.(fn{1});    % 最初の変数を取得

% mat は 3×N または N×3 を仮定
if size(mat,1)>=3 && size(mat,2)>3
  % 3行： time; y; u
  time = mat(1,:).';
  y    = mat(2,:).';
  u    = mat(3,:).';
elseif size(mat,2)>=3
  % 3列： [time,y,u]
  time = mat(:,1);
  y    = mat(:,2);
  u    = mat(:,3);
else
  error('I/O データのサイズが想定外です');
end

% dt の計算
if numel(time)>1
  dtv = diff(time);
  dt  = mean(dtv);
  if max(abs(dtv-dt)) > 0.01*dt
    warning('Non-uniform time steps detected.');
  end
else
  dt = Ts;
end
fprintf('Loaded I/O: %d samples, dt = %.4g s\n',numel(time),dt);

% 必要なら再サンプリング
if abs(dt-Ts)>1e-6
  fprintf('Resampling to FIR Ts...\n');
  t_old = time;
  t_new = (0:ceil(time(end)/Ts)).'*Ts;
  u = interp1(t_old,u,t_new,'linear','extrap');
  y = interp1(t_old,y,t_new,'linear','extrap');
  dt = Ts;
end
N = numel(u);

%% 4) RLS 初期化
%% 4) Partial‐Update LMS initialisation
h     = h_init(:);           % FIR taps
mu    = 1e-3;                % LMS stepsize (tune to your data)
M     = 10;                  % number of taps to update per sample
phi   = zeros(L,1);

yhat  = zeros(N,1);
err   = zeros(N,1);

%% 5) Real‐time loop with M‐Max partial‐update LMS
figure('Name','Partial‐Update LMS Identification');
ax1 = subplot(2,1,1); hold on; grid on;
h_meas = plot(NaN,NaN,'k');
h_pred = plot(NaN,NaN,'r--');
legend('Measured','Predicted');
xlabel('n'); ylabel('y');

ax2 = subplot(2,1,2); hold on; grid on;
h_err = plot(NaN,NaN,'b');
xlabel('n'); ylabel('error');

ax3 = subplot(3,1,3); 
h_coef = plot(1:L, h, 'b-o', 'MarkerSize',3); 
grid on;
xlabel('Tap index');
ylabel('h');
title(ax3,'Iteration 0');
ylim([min(h_init)*1.1, max(h_init)*1.1]); % 適宜レンジ調整

for n = 1:N
  % update input vector
  phi = [u(n); phi(1:end-1)];
  
  % prediction and error
  yhat(n) = phi.' * h;
  err(n)  = y(n) - yhat(n);
  
  % partial‐update LMS
  delta = mu * phi * err(n);
  [~, idx_sort] = sort(abs(delta), 'descend');
  S = idx_sort(1:min(M, L));
  h(S) = h(S) + delta(S);

  % store coefficient history
  H(:,n) = h;

  % update plots periodically
  if mod(n, plot_rate) == 0 || n == N
    set(h_meas, 'XData', 1:n, 'YData', y(1:n));
    set(h_pred, 'XData', 1:n, 'YData', yhat(1:n));
    set(h_err , 'XData', 1:n, 'YData', err(1:n));
    set(h_coef, 'YData', h);
    title(ax3, sprintf('Iteration %d', n));
    drawnow limitrate;
  end
end

%% 6) Error metrics & save (unchanged)
rmse  = sqrt(mean(err(L+1:end).^2));
yn    = y(L+1:end)-mean(y(L+1:end));
nrmse = 1 - norm(err(L+1:end))/norm(yn);
R2    = 1 - sum(err(L+1:end).^2)/sum(yn.^2);

fprintf('\n===== FINAL ERROR =====\n');
fprintf('RMSE  = %.4g\n',rmse);
fprintf('NRMSE = %.2f %%\n',nrmse*100);
fprintf('R^2   = %.3f\n',R2);

save('pu_lms_results.mat','h','rmse','nrmse','R2','yhat','err','Ts');
